doc/pserver.md

```diff
@@ -0,0 +1,226 @@
+# Parameter Server Design
+
+
+## Overview
+
+
+## PServer
+
+
+### Compoments
+
+
+![pserver](./images/pserver.png)
+
+
+PServer contains three main compoments:
+
+- KVStore
+- Gradient Queue
+- Optimizer
+
+1. The worker initializes a model, and pushes parameters to KVStore.
+2. Before each step of training, the worker pulls the latest model from KVStore.
+3. After forward/backward computation, the worker push gradients to Gradient Queue waiting for processing.
+4. Optimizer gets a gradient from the Gradient Queue.
+5. Then, Optimizer looks up the corresponding parameter from KVStore.
+6. Optimizer applies gradients to parameters, and updates parameter back to KVStore.
+
+
+The interfaces of KVStore could be like this:
+
+
+```python
+class KVStore(elasticdl_pb2_grpc.KVStoreServicer):
+    def __init__(self):
+        self.param_db = {}
+        self.embedding_param_db = {}
+
+    def pull_model(self, request, _):
+        pass
+
+    def push_model(self, request, _):
+        pass
+    
+    # embedding param is handled lazily   
+    def pull_embedding_param(self, reques, _):
+        pass
+
+    def get_param(self):
+        pass
+
+    def set_param(self):
+        pass
+```
+
+
+The interface of Gradient Queue could be like this:
+
+```python
+class GradientQueue(elasticdl_pb2_grpc.GradientQueueServicer):
+    def __init__(self):
+        self.grad_queue = queue.Queue()
+        
+    def push_gradient(self, request, _):
+        pass
+    
+    def get_gradient(self):
+        pass
+        
+    def put_gradient(self):
+        pass
+```
+
+### Parameters Transfromation
```

Comment:
We need to add a subsection to describe the data structure in PS for storing model variables and embedding vectors.
For a model variable, the key is the variable name, and the value is `tf.Variable`.
For an embedding vector, the key is embedding layer name + id, the value is a Numpy array.